src/py_zfs_vdev.c

```diff
@@ -0,0 +1,320 @@
+		    &child, &children) == 0);
+		Py_END_ALLOW_THREADS
+		if (found) {
+			for (uint_t i = 0; i < children; ++i) {
+				vdev_get_disks_impl(list, child[i]);
+			}
+		}
+	}
+	return;
+}
+
+static
+PyObject *py_zfs_vdev_get_disks(PyObject *self, PyObject *args) {
+	py_zfs_vdev_t *v = (py_zfs_vdev_t *)self;
+	PyObject *list;
+
+	list = PyList_New(0);
+	PYZFS_ASSERT(list, "Failed to create Python list object.");
+	vdev_get_disks_impl(list, v->vdev_tree);
+
+	return (list);
+}
+
+static
+PyGetSetDef zfs_vdev_getsetters[] = {
+	{
+		.name	= "type",
+		.get	= (getter)py_zfs_vdev_get_type
+	},
+	{
+		.name	= "path",
+		.get	= (getter)py_zfs_vdev_get_path
+	},
+	{ .name = NULL }
+};
+
+static
+PyMethodDef zfs_vdev_methods[] = {
+	{
+		.ml_name = "asdict",
+		.ml_meth = py_zfs_vdev_asdict,
+		.ml_flags = METH_VARARGS | METH_KEYWORDS
+	},
+	{
+		.ml_name = "name",
+		.ml_meth = py_zfs_vdev_get_name,
+		.ml_flags = METH_NOARGS,
+	},
+	{
+		.ml_name = "guid",
+		.ml_meth = py_zfs_vdev_get_guid,
+		.ml_flags = METH_NOARGS,
+	},
+	{
+		.ml_name = "status",
+		.ml_meth = py_zfs_vdev_get_status,
+		.ml_flags = METH_NOARGS,
+	},
+	{
+		.ml_name = "size",
+		.ml_meth = py_zfs_vdev_get_size,
+		.ml_flags = METH_NOARGS,
+	},
+	{
+		.ml_name = "children",
+		.ml_meth = py_zfs_vdev_get_children,
+		.ml_flags = METH_NOARGS,
+	},
+	{
+		.ml_name = "disks",
+		.ml_meth = py_zfs_vdev_get_disks,
+		.ml_flags = METH_NOARGS,
+	},
+	{ NULL, NULL, 0, NULL }
+};
+
+PyTypeObject ZFSVdev = {
+	.tp_name = "ZFSVdev",
+	.tp_basicsize = sizeof (py_zfs_vdev_t),
+	.tp_methods = zfs_vdev_methods,
+	.tp_getset = zfs_vdev_getsetters,
+	.tp_new = py_zfs_vdev_new,
+	.tp_init = py_zfs_vdev_init,
+	.tp_doc = "ZFSVdev",
+	.tp_dealloc = (destructor)py_zfs_vdev_dealloc,
+	.tp_repr = py_repr_zfs_vdev,
+	.tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE
+};
+
+py_zfs_vdev_t *init_zfs_vdev(py_zfs_t *lzp, zpool_handle_t *zhp,
+    nvlist_t *tree, PyObject *parent)
+{
+	py_zfs_vdev_t *out = NULL;
+	const char *type;
+	const char *path;
+	boolean_t path_found;
+
+	out = (py_zfs_vdev_t *)PyObject_CallFunction((PyObject *)&ZFSVdev, NULL);
+	PYZFS_ASSERT(out, "Failed to create ZFSVdev type object.");
+
+	out->pylibzfsp = lzp;
+	Py_INCREF(lzp);
+	out->parent = parent;
+	Py_XINCREF(parent);
+	out->zhp = zhp;
+	out->vdev_tree = tree;
+
+	Py_BEGIN_ALLOW_THREADS
+	type = fnvlist_lookup_string(tree, ZPOOL_CONFIG_TYPE);
+	path_found = nvlist_lookup_string(tree, ZPOOL_CONFIG_PATH, &path) == 0;
+	Py_END_ALLOW_THREADS
+
+	if (strncmp(type, VDEV_TYPE_RAIDZ, strlen(VDEV_TYPE_RAIDZ)) == 0) {
+		int parity;
+		char ftype[20];
+
+		Py_BEGIN_ALLOW_THREADS
+		parity = fnvlist_lookup_uint64(tree, ZPOOL_CONFIG_NPARITY);
+		Py_END_ALLOW_THREADS
+
+		snprintf(ftype, sizeof(ftype), "%s%d", type, parity);
+		out->type = PyUnicode_FromString(ftype);
+	} else
+		out->type = PyUnicode_FromString(type);
+	PYZFS_ASSERT(out->type, "Failed to create vdev type string.");
```

Comment:
Can we `Py_DECREF(out);` and return NULL instead?

Reply:
Updated to return NULL here.